src/jit/decomposelongs.cpp

```diff
@@ -964,20 +975,252 @@ GenTree* DecomposeLongs::DecomposeShift(LIR::Use& use)
             unreached();
     }
 
-    GenTreeArgList* argList = m_compiler->gtNewArgList(loOp1, hiOp1, shiftWidthOp);
+    // If we are shifting by a constant int, we do not want to save to a helper, instead, we decompose.
+    if (shiftOper == GT_CNS_INT)
+    {
+        GenTree* loCopy  = m_compiler->gtNewLclvNode(loOp1LclNum, TYP_INT);
 
-    GenTree* call = m_compiler->gtNewHelperCallNode(helper, TYP_LONG, 0, argList);
+        GenTree* loResult;
+        GenTree* hiResult;
 
-    GenTreeCall*    callNode    = call->AsCall();
-    ReturnTypeDesc* retTypeDesc = callNode->GetReturnTypeDesc();
-    retTypeDesc->InitializeLongReturnType(m_compiler);
+        switch (oper)
+        {
+            case GT_LSH:
+            {
+                if (count == 0)
+                {
+                    // Do nothing.
+                    loResult = loOp1;
+                    hiResult = hiOp1;
+                    Range().InsertAfter(tree, loResult, hiResult);
+                }
+                else if (count < 32)
+                {
+                    // Hi is a GT_LSH_HI, lo is a GT_LSH. Will produce:
+                    // reg1 = lo
+                    // shl lo, shift
+                    // shld hi, reg1, shift
+
+                    GenTree* shiftByHi = m_compiler->gtNewIconNode(count, TYP_INT);
+                    GenTree* shiftByLo = m_compiler->gtNewIconNode(count, TYP_INT);
+
+                    loResult = m_compiler->gtNewOperNode(GT_LSH, TYP_INT, loOp1, shiftByLo);
+
+                    // Create a GT_LONG that contains loCopy and hiOp1. This will be used in codegen to
+                    // generate the shld instruction
+                    GenTree* hiOp = new (m_compiler, GT_LONG) GenTreeOp(GT_LONG, TYP_LONG, loCopy, hiOp1);
+                    hiResult      = m_compiler->gtNewOperNode(GT_LSH_HI, TYP_INT, hiOp, shiftByHi);
+
+                    Range().InsertAfter(tree, loOp1, shiftByLo, loResult);
+                    Range().InsertAfter(loResult, loCopy, hiOp1, hiOp);
+                    Range().InsertAfter(hiOp, shiftByHi, hiResult);
+                }
```

Comment:
As a general point of style, one thing you _could_ do (certainly not necessary) is convert all the inserts to InsertBefore(), in which case you could probably always use `tree` as the insert point, instead of needing to change the insert point as you add additional insertions. This is maybe more useful in the >=32 case below, where the insertions are not all together.